### PR TITLE
Create pre-locked key range FT cursors for some index scans.

### PR DESCRIPTION
**Pull request by christianrober**
These changes allow the fractal tree storage engine to create cursors that pre-lock a range of keys before they are used/consumed/altered by a write operation. In Read-Modify-Write workloads, like db.collection.update() operations, this will help reduce lock contention and improve concurrency.

The high level goal of pre-locking the cursor range is to lock and serialize write access to logical keys before descending deep into the fractal tree, where even more locks will be held. By avoiding this nested locking, other operations will be blocked less by a concurrent write operation.

Note: Changes to the MongoDB source code were required to enable pre-locking. These primarily include a minor refactoring of index scan initialization and some new key range caching methods added to the operation context classes.

### CODE REVIEW

**Comment by dbpercona** [comment]
Regarding the MAX constants added in toku_dictionary.cpp,  this should probably be defined once in toku_dictionary.h with a more descriptive name and use it in both methods.   In the future we may decide to make it a configurable parameter.

File: src/mongo/db/storage/tokuft/tokuft_dictionary.cpp [hunk 1]
```diff
@@ -202,6 +231,72 @@ namespace mongo {
         }
     }
 
+    KVDictionary::Cursor *TokuFTDictionary::CreatePrelockedCursorWithRetryAndStartKey(OperationContext *opCtx,
+                                                                                      const Slice &key,
+                                                                                      const int direction) const {
+        const int MAX = 32; // <-- This is a rough guess/maybe maps to threads.
```
Comment (dbpercona) [inline src/mongo/db/storage/tokuft/tokuft_dictionary.cpp]: Regarding the MAX constants added in toku_dictionary.cpp, this should probably be defined once in toku_dictionary.h with a more descriptive name and use it in both methods. In the future we may decide to make it a configurable parameter.

**Comment by dbpercona** [comment]
Outside of the one issue with the contants in toku_dictionary.cpp this looks good!

File: src/mongo/db/storage/tokuft/tokuft_dictionary.cpp (hunk 1 above)
Comment (christianrober) [inline src/mongo/db/storage/tokuft/tokuft_dictionary.cpp]: I agree, this name should be better.

**Pull request merged**